024-revise-operations.md

```diff
@@ -0,0 +1,521 @@
+been documented in the past.
+
+Proposed Changes
+----------------
+
+In Chapter 3 replace the following paragraph:
+
+> A DWARF expression is encoded as a stream of operations, each
+> consisting of an opcode followed by zero or more literal operands. The
+> number of operands is implied by the opcode.
+
+With:
+
+> A DWARF expression is encoded as a stream of operations, each
+> operation consisting of an opcode followed by zero or more inline
+> parameters. The number of inline parameters is implied by the
+> opcode. It may also receive operands from the stack and make use of
+> information from its evaluation context.
+
+>
+> *The structure description of the inputs and each of each operation
+> will be as follows:*
+> `DW_OP_name` ([*type*] first inline operand>, [*type*] second inline operand, ...)
+>
+>     <[*type*] first stack argument> <[*type*>] second stack argument> ...
+>         → <[*type*] first stack result> <[*type*] second stack result> ...
+>
+
+> *The first line mentions the inline operands and their expected
+> types or binary encodings. These encodings can be found in Section X.Y*
+
+> *The second line explains the ways that the operator affects the
+> DWARF stack. The arguments consumed from the stack are listed
+> first. The stack is built with new entries added to the right. The
+> top of the stack is on the right while the deepest entry is on the
+> stack is on the left. Thus: "push A" followed by "push B" would yield:*
+
+>    < A> < B> 
+
+> *Then after the "→" the results that are pushed on the stack. Each
+> entry on the DWARF stack has a type. In most cases, it is broadly a
+> value or location. However, there are often limits placed on the
+> each operator's input stack arguments limiting the sub-types over
+> which the operator is defined. Since one operator's output becomes
+> the input for subsequent operators in a DWARF expression, the type
+> of the operation's output is also often listed when it cannot be
+> immediately inferred.*
+
+> *Symbolic names are often given to the stack arguments and the stack
+> result when it is considered to be helpful or meaningful. In the
+> cases where no sybolic name would be meaningful, the generic names A
+> B C and D are used. When, an operator modifies a stack argument in
+> meaningful way before returning it to the stack, the name is given
+> an apostrophe suffix, for example turning A into A' ("A prime").*
+
+In Section 3.2
+
+> `DW_OP_dup`
+>
+>       <[*any*] A> → < A> < A>
+>
+
+> `DW_OP_drop`
+>
+>       <[*any*] A> <[*any*] B>  → < B>
+>
+
+> `DW_OP_pick` ([1-byte integral] N)
+>
+>      <[*any*] Nth>  ... < top> → < Nth> ... < top> < Nth>
+>
+
+**NOTE FOR DISCUSSION**: It should be noted that there is no pick
+operation which can pick based upon an index which is passed on the
+stack instead of as a literal operand. The fact that it hasn't been
+requested may suggest that it is not needed.
+
+> `DW_OP_over`
+>
+>     <[*any*] A> <[*any*] B> → < B> < A> < B>
```

Comment:
Shouldn't the result be `<A> <B> <A>`?  L55-61 explains that the top of stack is on the right.